devicetypes/ezex/somfy-blind-controller-ezex.src/somfy-blind-controller-ezex.groovy

```diff
@@ -0,0 +1,120 @@
+              attributeState "turningOff", label:'Closing', icon:"http://www.ezex.co.kr/img/st/window_close.png", backgroundColor:"#ffffff", nextState:"turningOn"  
+              //attributeState "partially open", label:'Partially open', action: "switch.off", icon:"http://www.ezex.co.kr/img/st/window_open.png", backgroundColor:"#d45614", nextState: "turningOff"
+            }         
+        }
+        standardTile("home", "device.level", width: 2, height: 2, decoration: "flat") {
+          state "default", label: "home", action:"presetPosition", icon:"st.Home.home2"
+        }
+        standardTile("contPause", "device.switch", inactiveLabel: false, decoration: "flat", width: 2, height: 2) {
+        	state "pause", label:"", icon:'st.sonos.pause-btn', action:'pause', backgroundColor:"#cccccc"         
+        }
+        standardTile("refresh", "device.refresh", inactiveLabel: false, decoration: "flat", width: 2, height: 1) {
+          state "default", label:"", action:"refresh.refresh", icon:"st.secondary.refresh"
+        }      
+        valueTile("ShadeLevel", "device.level", width: 4, height: 1) {
+        	state "level", label: 'Shade is ${currentValue}% up', defaultState: true
+        }
+        
+        controlTile("levelSliderControl", "device.level", "slider", width:2, height: 1, inactiveLabel: false) {
+        	state "level", action:"switch level.setLevel"
+        }
+        
+        preferences {
+          input "preset", "number", title: "Default half-open position (1-100). Springs Window Fashions users should consult their manuals.", defaultValue: 50, required: false, displayDuringSetup: false
+        }
+      
+        main "windowShade"
+        details(["windowShade", "contPause", "ShadeLevel", "levelSliderControl","refresh"])
+    }
+}
+
+// Parse incoming device messages to generate events
+def parse(String description) {
+
+    
+    def parseMap = zigbee.parseDescriptionAsMap(description)
+    def event = zigbee.getEvent(description)
+    
+    def attrs = parseMap.additionalAttrs
+    if( attrs != null ) { 
+				attrs.each { attr ->
+            if(attr.value == "64") { //open
+                sendEvent(name: "windowShade", value: "open")
+                sendEvent(name: "level", value: "100")
+            }
+            else if(attr.value == "00"){ //closed
+                sendEvent(name: "windowShade", value: "closed")
+                sendEvent(name: "level", value: "0")
+            } else {           	
+                sendEvent(name: "windowShade", value: "partially open")
+                sendEvent(name: "level", value: zigbee.convertHexToInt(attr.value))
+            }
+        }
+     }
+}
+
+def close() {
+	  log.info "LOG-MH-Shade: off()"
+      zigbee.command(0x0102, 0x01)
+}
+
+def open() {
+	  log.info "LOG-MH-Shade: on()"
+      zigbee.command(0x0102, 0x00)
+}
+
+def setLevel(data) {
```

Comment:
add `rate=null` to the signature here

Reply:
Could you explain in detail why it should have `rate=null` even though this setLevel() is a local method in this DTH. If it also needs to treat rate in the body I could guess something but just adding to the signature didn't make sense to me.

Reply:
`setLevel` is always called with two parameters when executed via the public API and other users of the command may choose to pass `rate` or not. Basically it's optional for the caller but required to be implemented. We recently updated all of our existing DTHs to add the `rate` parameter where it was missing for this reason.

Reply:
When I searched many DTHS in the public, I just could find one.
Anyway should I implement the body treating `rate` , I could copy it from appengine zigbee lib. Will it be OK?

Reply:
This is where we changed `setLevel` in  existing DTHs: https://github.com/SmartThingsCommunity/SmartThingsPublic/pull/3999. 

For the implementation just define `setLevel` as `def setLevel(value, rate = null)` and don't do anything with the `rate` parameter since the  `Go To Lift Percentage` command doesn't take a rate parameter.

Reply:
Thanks for the reference, I could understand it clearly. I didn't realize there has been the `rate `in the Switch Level.  I will apply it.